tests/test_inputs_handler_matsim_reader.py

```diff
@@ -338,3 +340,23 @@ def test_reading_pt2matsim_vehicles():
                 'egressTime': {'secondsPerPerson': '0.5'},
                 'doorOperation': {'mode': 'serial'},
                 'passengerCarEquivalents': {'pce': '2.8'}}})
+
+
+def test_reading_HK_network_with_elevation():
+    n = read.read_matsim(path_to_network=HK_network, epsg='epsg:4326')
+
+    assert_semantically_equal(dict(n.nodes()), {
+        '101982': {'id': '101982', 'x': 114.161432, 'y': 22.279784, 'z': 25.0, 'lon': 114.161432, 'lat': 22.279784,
+                    's2_id': 3748121220106005759},
+         '101986': {'id': '101986', 'x': 114.159648, 'y': 22.278037, 'z': 53.0, 'lon': 114.159648, 'lat': 22.278037,
+                    's2_id': 3748121226099361651}})
+
+    assert_semantically_equal(dict(n.links()), {
+        '0': {'id': '0', 'from': '101982', 'to': '101986', 'freespeed': 4.166666666666667, 'capacity': 600.0,
+               'permlanes': 1.0, 'oneway': '1', 'modes': {'car'}, 's2_from': 3748121220106005759,
```

Comment:
Again we have a lot of magic numbers - can we be clearer in the test about why we expect these?

Reply:
These are just components of the network xml, I think this is the only option if we are trying to test the function that reads the xml - a bit of a chicken and egg problem again it seems to me

Reply:
I wouldn't say it's a chicken and egg problem. If I understand you correctly, what you're talking about is the possibility of recreating all of the network parsing logic in the test, but that is not what I'm talking about, and it is not our only option if our intention is to make the test clearer. What we're trying to achieve is to **reduce or eliminate the disconnect between our test data and the expectations we use in our assertions**.

One thing you could do is create the XML network file from a string in your test (or fixture), for example. That way, the network and its various values are right there in your test, so it becomes obvious what our expectation should be. You could use helper methods that fill in all the boilerplate, and you just pass in the values you will later assert on, like a list of nodes and links and their properties, say.

In this case, as far as I can see, that test data file is only being used in this one place anyway, plus it's quite small, so I don't think you gain anything much by having it live outside the test:

<img width="737" alt="Screen Shot 2022-04-20 at 13 38 06" src="https://user-images.githubusercontent.com/250899/164232242-cfacc901-e0ea-49ec-88b9-beb01df9faf9.png">

I can rewrite the test in this manner if it will help you to visualise it.